### PR TITLE
fix: Move service initialization outside of plugins

### PR DESCRIPTION
**Pull request by tbarlow12**
Fixes error we found in testing serverless optional variables (`${opt:stage, 'dev'}`), which allow you to specify a param at deploy time, as well as a default value. Also fixes potentially more serious problems with setting names before the configuration is initialized.

### CODE REVIEW

**Comment by mydiemho** [comment]
> Fixes error we found in testing serverless optional variables (`${opt:stage, 'dev'}`), which allow you to specify a param at deploy time, as well as a default value. Also fixes potentially more serious problems with setting names before the configuration is initialized.

1. what is the error you see when testing ` (`${opt:stage, 'dev'}`),`?
1. What are the serious problems with setting names before configuration is initialized?

File: src/services/baseService.ts [hunk 1]
```diff
@@ -256,8 +256,13 @@ export abstract class BaseService {
    * Overwrite values for resourceGroup, prefix, region and stage
    * in config if passed through CLI
    */
-  private setConfigFromCli() {
+  private setupConfig() {
     const { prefix, region, stage, subscriptionId } = this.config.provider;
+
+    if (stage.startsWith("${")) {
```
Comment (mydiemho) [inline src/services/baseService.ts]: why is this needed?

File: src/plugins/offline/azureOfflinePlugin.ts [hunk 2]
```diff
@@ -60,14 +58,17 @@ export class AzureOfflinePlugin extends AzureBasePlugin {
   }
 
   private async azureOfflineBuild(){
-    await this.offlineService.build();
+    const offlineService = new OfflineService(this.serverless, this.options);
```
Comment (mydiemho) [inline src/plugins/offline/azureOfflinePlugin.ts]: probably not too much of a difference, but what are we given up creating a new service everytime?

File: src/services/baseService.ts (hunk 1 above)
Comment (tbarlow12) [reply src/services/baseService.ts]: Leftover from before. Good callout :)

File: src/plugins/offline/azureOfflinePlugin.ts (hunk 2 above)
Comment (tbarlow12) [reply src/plugins/offline/azureOfflinePlugin.ts]: We can't initialize a `BaseService` within the constructor of a plugin. The constructor of `BaseService` assumes that the core framework has made whatever config substitutions it needs to do

Comment (wbreza) [reply src/plugins/offline/azureOfflinePlugin.ts]: This looks good enough for now - but I feel the `BaseService` is starting to do too much and we should consider scaling it back at some point.  Ex) It does a lot of config initialization and if you think about all the different services we instantiate in the course of calling a plugin the same code is getting run over and over.
